sdk/core/core/src/az_http_policy_retry.c

```diff
@@ -0,0 +1,228 @@
+}
+
+AZ_INLINE az_result _az_http_policy_retry_append_http_retry_msg(
+    int16_t attempt,
+    int32_t delay_msec,
+    az_span* ref_log_msg)
+{
+  AZ_RETURN_IF_FAILED(
+      az_span_append(*ref_log_msg, AZ_SPAN_FROM_STR("HTTP Retry attempt #"), ref_log_msg));
+
+  AZ_RETURN_IF_FAILED(az_span_append_i32toa(*ref_log_msg, (int32_t)attempt, ref_log_msg));
+
+  AZ_RETURN_IF_FAILED(
+      az_span_append(*ref_log_msg, AZ_SPAN_FROM_STR(" will be made in "), ref_log_msg));
+
+  AZ_RETURN_IF_FAILED(az_span_append_i32toa(*ref_log_msg, delay_msec, ref_log_msg));
+  AZ_RETURN_IF_FAILED(az_span_append(*ref_log_msg, AZ_SPAN_FROM_STR("ms."), ref_log_msg));
+
+  return AZ_OK;
+}
+
+AZ_INLINE void _az_http_policy_retry_log(int16_t attempt, int32_t delay_msec)
+{
+  uint8_t log_msg_buf[AZ_LOG_MSG_BUF_SIZE] = { 0 };
+  az_span log_msg = AZ_SPAN_FROM_BUFFER(log_msg_buf);
+  (void)_az_http_policy_retry_append_http_retry_msg(attempt, delay_msec, &log_msg);
+  az_log_write(AZ_LOG_HTTP_RETRY, log_msg);
+}
+
+AZ_INLINE AZ_NODISCARD int32_t _az_uint32_span_to_int32(az_span span)
+{
+  uint32_t value = 0;
+  az_result const itoa_result = az_span_to_uint32(span, &value);
+  bool const itoa_succeeded = az_succeeded(itoa_result);
+  if (itoa_result == AZ_ERROR_ARITHMETIC_OVERFLOW || (itoa_succeeded && value > INT32_MAX))
+  {
+    return INT32_MAX;
+  }
+  else if (itoa_succeeded)
+  {
+    return value;
+  }
+
+  return -1;
+}
+
+AZ_INLINE AZ_NODISCARD az_result _az_http_policy_retry_get_retry_after(
+    az_http_response* ref_response,
+    az_http_status_code const* status_codes,
+    bool* should_retry,
+    int32_t* retry_after_msec)
+{
+  az_http_response_status_line status_line = { 0 };
+  AZ_RETURN_IF_FAILED(az_http_response_get_status_line(ref_response, &status_line));
+  az_http_status_code const response_code = status_line.status_code;
+
+  for (; *status_codes != AZ_HTTP_STATUS_CODE_NONE; ++status_codes)
+  {
+    if (*status_codes != response_code)
+    {
+      continue;
+    }
+
+    // Try to get the value of retry-after header, if there's one.
+    *should_retry = true;
+
+    az_pair header = { 0 };
+    while (az_http_response_get_next_header(ref_response, &header) == AZ_OK)
+    {
+      if (az_span_is_content_equal_ignoring_case(header.key, AZ_SPAN_FROM_STR("retry-after-ms"))
+          || az_span_is_content_equal_ignoring_case(
+              header.key, AZ_SPAN_FROM_STR("x-ms-retry-after-ms")))
+      {
+        // The value is in milliseconds.
+        int32_t const msec = _az_uint32_span_to_int32(header.value);
+        if (msec >= 0) // int32_t max == ~24 days
+        {
+          *retry_after_msec = msec;
+          return AZ_OK;
+        }
+      }
+      else if (az_span_is_content_equal_ignoring_case(header.key, AZ_SPAN_FROM_STR("Retry-After")))
+      {
+        // The vaule is either seconds or date.
+        int32_t const seconds = _az_uint32_span_to_int32(header.value);
+        if (seconds >= 0) // int32_t max == ~68 years
```

Comment:
What do we do when seconds is negative because we have "invalid" header.value? It looks like we'd return `AZ_OK`. Is that what we want?

Reply:
Yes - we would initialize retry_after_msec with "-1" indicating that we didn't manage to get the value from the "Retry-After" headers, which is still AZ_OK. In that case, we'll fall back to the default behavior which is to wait exponentially.

Reply:
I am consider this PR feedback as 'done' for the scope of this PR, and not going to do more commits,and I'm going to resolve this conversation. Please revert it or ping me if you disagree and think that I should not check this in.